rust/src/flatfit/mod.rs

```diff
@@ -0,0 +1,129 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+
+const LOW_CAP: usize = 2;
+
+#[derive(Clone)]
+pub struct FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    front: usize,
+    back: usize,
+    size: usize,
+    buffer: RefCell<Vec<Item<Value>>>,
+    tracing_indices: RefCell<Vec<usize>>,
+    binop: PhantomData<BinOp>,
+}
+
+#[derive(Clone)]
+struct Item<Value> {
+    val: Value,
+    next: usize,
+}
+
+impl<Value> Item<Value> {
+    fn new(val: Value, next: usize) -> Self {
+        Self { val, next }
+    }
+}
+
+impl<Value, BinOp> FifoWindow<Value, BinOp> for FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new() -> Self {
+        Self {
+            front: 0,
+            back: 0,
+            size: 0,
+            buffer: RefCell::new(Vec::new()),
+            tracing_indices: RefCell::new(Vec::new()),
+            binop: PhantomData,
+        }
+    }
+    fn push(&mut self, val: Value) {
+        let capacity = self.buffer.borrow().capacity();
+        if self.size + 1 > capacity {
+            self.rescale(capacity * 2)
+        }
+        let mut buffer = self.buffer.borrow_mut();
+        self.size += 1;
+        let prev = self.back;
+        self.back = (self.back + 1) % self.size;
+        buffer[self.back].val = val;
+        buffer[prev].next = self.back;
+    }
+    fn pop(&mut self) {
+        let capacity = self.buffer.borrow().capacity();
+        if self.size > 0 {
+            self.front = (self.front + 1) % capacity;
+            self.size -= 1;
+            if self.size < capacity / 2 {
+                self.rescale(capacity / 2)
+            }
+        }
+    }
+    fn query(&self) -> Value {
+        let mut agg = Value::identity();
+        let mut buffer = self.buffer.borrow_mut();
+        if self.size > 0 {
+            let mut tracing_indices = self.tracing_indices.borrow_mut();
+            tracing_indices.clear();
```

Comment:
Maybe an `assert!(tracing_indices.is_empty())` is better. All of the pops from the previous query _should_ have completely cleared out the vector. If `tracing_indices` is not empty at this point, I believe that would violate invariants.